File: DSA/SakshamDwivedi_590013147/Day17/D17P2.py
#Solution Function
def first_occ_bin(arr, target):
    start = 0
    end = len(arr) - 1
    current_index = len(arr)
    while start <= end:
        mid = start + (end - start) // 2
        if arr[mid] > target:
            end = mid - 1
        elif arr[mid] < target:
            start = mid + 1
        elif ((arr[mid] == target) and (current_index > mid)):
            current_index = mid
            end = mid - 1
    if(current_index < len(arr)):
        return current_index
    else:
        return -1

#Fixed Input
array = [1,1,1,2,3,3]
tar = 1
print(first_occ_bin(array, tar))

#Custom Input
array = list(int(x) for x in input("Enter an array : ").split())
tar = int(input("Enter the target value : "))
print(first_occ_bin(array, tar))

File: DSA/SakshamDwivedi_590013147/Day17/D17P2.c
#include <stdio.h>

int arr[100];
int tar;
int size;

int main(){
    printf("Enter the size of the array (Less than 100): ");
    scanf("%d", &size);
    printf("Enter the elements (In sorted order) : ");
    for(int i = 0; i < size; i++){
        scanf("%d", &arr[i]);
    }
    printf("Enter the target value to search : ");
    scanf("%d", &tar);
    int start = 0;
    int end = size - 1;
    int current_index = size;
    while(start <= end){
        int mid = start + (end - start) / 2;
        if(arr[mid] > tar){
            end = mid - 1;
        }
        else if(arr[mid] < tar){
            start = mid + 1;
        }
        else if((arr[mid] == tar) && (current_index > mid)){
            current_index = mid;
            end = mid - 1;
        }
    }
    if(current_index < size){
        printf("%d", current_index);
    }
    else{
        printf("-1");
    }
}
